File: 454_4Sum_II.py
class Solution:
    def fourSumCount(self, A: List[int], B: List[int], C: List[int], D: List[int]) -> int:
        dic = {}
        count = 0
        for a in A:
            for b in B:
                s = a + b
                dic[s] = dic.get(s, 0) + 1     
        for c in C:
            for d in D:
                s = c + d
                count += dic.get(-s, 0)
        return count

File: 454_4Sum_II.cpp
class Solution {
public:
    int fourSumCount(vector<int>& A, vector<int>& B, vector<int>& C, vector<int>& D) {
        unordered_map<int, int> AB;
        int res = 0;
        for (int a : A) {
            for (int b : B) {
                AB[a + b] += 1;
            }
        }
        
        for (int c : C) {
            for (int d : D) {
                res += AB[-c - d];
            }
        }
        
        return res;
    }
};
